docs/extend/api.md

```diff
@@ -12,348 +12,967 @@ To use the built-in REST API as part of an integration, see [Consuming the REST
+
+```php
+/** @var \Flarum\Http\UrlGenerator $url */
+$url->to('api')->route('labels.index');
+$url->to('api')->route('labels.activate', ['id' => $label->id]);
 ```
 
-#### Searching and Filtering
+## Fields (Attributes and Relationships)
 
-Read our [searching and filtering](search.md) guide for more information!
+The `fields` method on the API resource is used to define the fields (attributes and relationships) of the model. You can use the `Schema` namespace to define the various field types.
 
-### Showing a Resource
+### Attributes
 
-For the controller that shows a single resource, extend the `Flarum\Api\Controller\AbstractShowController` class. Like for the list controller, you need to specify the `$serializer` you want to use to serialize your models, and implement a `data` method to return a single model. We'll learn about serializers [in just a bit](#serializers).
+Before you define an attribute, decide which type of attribute it is. The `Schema` namespace provides a range of attribute types, such as `Str`, `Integer`, `Boolean`, `DateTime` and `Arr` for arrays.
 
 ```php
-use Flarum\Api\Controller\AbstractShowController;
-use Illuminate\Support\Arr;
-use Psr\Http\Message\ServerRequestInterface as Request;
-use Tobscure\JsonApi\Document;
+use Flarum\Api\Schema;
 
-class ShowTagController extends AbstractShowController
+public function fields(): array
 {
-    public $serializer = TagSerializer::class;
-    
-    protected function data(Request $request, Document $document)
-    {
-        $id = Arr::get($request->getQueryParams(), 'id');
-        
-        return Tag::findOrFail($id);
-    }
+    return [
+        Schema\Str::make('name')
+            ->requiredOnCreate()
+            ->writable(),
+        Schema\Integer::make('discussionCount'),
+        Schema\Arr::make('customData'),
+        Schema\Boolean::make('isActive')
+            ->writable(),
+        Schema\DateTime::make('createdAt'),
+    ];
 }
 ```
 
-### Creating a Resource
+### Visibility
 
-For the controller that creates a resource, extend the `Flarum\Api\Controller\AbstractCreateController` class. This is the same as the show controller, except the response status code will automatically be set to `201 Created`. You can access the incoming JSON:API document body via `$request->getParsedBody()`:
+You can use the `visible` method to determine whether an attribute should be visible in the API response.
 
 ```php
-use Flarum\Api\Controller\AbstractCreateController;
-use Illuminate\Support\Arr;
-use Psr\Http\Message\ServerRequestInterface as Request;
-use Tobscure\JsonApi\Document;
+use Flarum\Api\Context;
+use Flarum\Api\Schema;
 
-class CreateTagController extends AbstractCreateController
+public function fields(): array
```

Comment:
maybe worth noting somewhere towards the top that all these examples are of methods in the resource class.